### PR TITLE
Fix display impl for empty enum and non provided enum

### PR DESCRIPTION
**Pull request by Mifom**
If `onlyerror::Error` is derived to an empty enum or enum with no `#[error]` macros and no doc comments on its variants the compile error is unclear:
```
1 | #[derive(Debug, onlyerror::Error)]
  |                 ^^^^^^^^^^^^^^^^ `Error` cannot be formatted with the default formatter
```
It happens because `core::error::Error` (and `std::error::Error`) trait requires `Display` to be implemented.

Fixed by adding unreachable branch to match and removing checking for any `display` provided.

Now, this code compiles:
```rust
#[derive(Debug, onlyerror::Error)]
enum Error {}
```

And this code:
```rust
#[derive(Debug, onlyerror::Error)]
enum Error {
    Some,
    Other,
}
```
produces error:
```
error: Required error message is missing
 --> src/main.rs:3:5
  |
3 |     Some,
  |
```

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -178,6 +178,9 @@ pub fn derive_error(input: TokenStream) -> TokenStream {
                 Ok(msg) => display_matches.push_str(&msg),
             }
         }
+        display_matches.push_str(&format!(
+            "_ => unsafe {{ ::{std_crate}::hint::unreachable_unchecked()}}"
```
Comment (parasyte) [inline src/lib.rs:179]: I would strongly prefer not outputting `unsafe`. Especially when we have nothing preventing the unreachable code from actually being unreachable.

For instance, this is trivially unsound:

```rust
#[derive(Debug, onlyerror::Error)]
enum Error {
    /// Some
    Some,
    Other,
}

fn main() {
    println!("{}", Error::Other);
}
```

**Review by parasyte — CHANGES_REQUESTED** [review]
There is definitely a bug here!

But I don't think this is the way to fix it. Looks like the way I wrote the code, I had intended to do the `display` verification for every enum variant, and return a `spanned_error()` with the span on the ident in https://github.com/parasyte/onlyerror/blob/a7edf5bf7933f5ec0b7b8b5bb79302ab5ba5721d/src/lib.rs#L137-L139

It's just incomplete error handling.

The reason for unsoundness with this PR (aside from the `unsafe`) is because the `ast.variants` list only contains the `Some` variant. It's missing `Other` because the parser is still dumb wrt. detecting values on unit variants. It inappropriately consumes an `Ident` which is intended to match the `crate` in:

```rust
enum Error {
    Value = crate::SOME_STATIC,
}
```

The broken parse code is here: https://github.com/parasyte/onlyerror/blob/a7edf5bf7933f5ec0b7b8b5bb79302ab5ba5721d/src/parser.rs#L124-L132

I'll work on this, and start to build out a lot more tests to handle these edge cases.

File: src/lib.rs (hunk 1 above)
Comment (Mifom) [reply src/lib.rs:179]: Thanks for the feedback! I've missed that case.

## About parsing
I've fixed parsing of unit type in 2 steps:
1. When checking on variant type the token is not consumed if it is not a group
2. If variant type is unit we can just skip until ',' as otherwise the syntax will not e correct

## About unsafe
The `unreachable_unchecked`branch is implemented as in std's derived impls for `Debug`, `Clone`, etc. I can replace it with `unreachable!()` if you prefer, this branch is obviously should be unreachable. It is here only because references are always considered inhabited.
If parsing step covers all enum variants(it should do that now) then there will be either compiler error for not providing display for any enum variant or all enum variants will be present in match.

Comment (Mifom) [reply src/lib.rs:179]: ## About tests
I don't know how to test code that should not compile. So if you have examples I would like to see and write tests for this PR.

Comment (parasyte) [reply src/lib.rs:179]: Thanks! I was working on the underlying parsing issue in https://github.com/parasyte/myn/pull/8 -- I think what you have here is good for now, and we can update `myn` separately.

[`trybuild`](https://crates.io/crates/trybuild) or [`ui_test`](https://crates.io/crates/ui_test) is the usual way to test "this does not compile" cases along with assertions over patterns in the error messages.

**Review by parasyte — COMMENTED** [review]
I don't have a whole lot of time to go over the PR right now, but I wanted to leave another comment. The reason for the empty-string case when there are no "display strings" is that I wanted it to be possible for callers to implement `Display` themselves. If we switch to a mode where `Display` is _always_ provided, then it becomes impossible for users to define their own.

I think that no matter what we do, this feature is going to require an additional annotation (perhaps `#[no_display]` or such) to make it work properly. (I've been thinking about this problem for a few days, and it's the best I've come up with given the permutations.)

Anyway, that feature can be deferred to another PR. I just wanted to point it out for context.

**Comment by Mifom** [comment]
> I don't have a whole lot of time to go over the PR right now, but I wanted to leave another comment. The reason for the empty-string case when there are no "display strings" is that I wanted it to be possible for callers to implement `Display` themselves. If we switch to a mode where `Display` is _always_ provided, then it becomes impossible for users to define their own.
> 
> I think that no matter what we do, this feature is going to require an additional annotation (perhaps `#[no_display]` or such) to make it work properly. (I've been thinking about this problem for a few days, and it's the best I've come up with given the permutations.)
> 
> Anyway, that feature can be deferred to another PR. I just wanted to point it out for context.

Thats a good idea. I've implemented `#[no_display]` attribute.

Also, added tests with `trybuild` for basic cases including `no_display`.

**Comment by parasyte** [comment]
I have one test failure on my end (using both stable and nightly). It appears to be some additional hint included in the error message. There are also a number of warnings from unused enum variants. Those should be silenced with `#![allow(dead_code)]`. And finally some clippy errors. I'll start the CI pipeline and you can see what it catches.

<details><summary>This is my patch so far to clean those up...</summary>

```diff
diff --git a/compile_tests/multiple_variant.rs b/compile_tests/multiple_variant.rs
index 9643058..1fad7f2 100644
--- a/compile_tests/multiple_variant.rs
+++ b/compile_tests/multiple_variant.rs
@@ -1,3 +1,5 @@
+#![allow(dead_code)]
+
 #[derive(Debug, onlyerror::Error)]
 enum Error {
     /// First
diff --git a/compile_tests/no_display.rs b/compile_tests/no_display.rs
index e680488..cf6952e 100644
--- a/compile_tests/no_display.rs
+++ b/compile_tests/no_display.rs
@@ -1,3 +1,5 @@
+#![allow(dead_code)]
+
 #[derive(Debug, onlyerror::Error)]
 #[no_display]
 enum Error {
diff --git a/compile_tests/no_display_no_impl.rs b/compile_tests/no_display_no_impl.rs
index 9f03f5e..b7af50f 100644
--- a/compile_tests/no_display_no_impl.rs
+++ b/compile_tests/no_display_no_impl.rs
@@ -1,3 +1,5 @@
+#![allow(dead_code)]
+
 #[derive(Debug, onlyerror::Error)]
 #[no_display]
 enum Error {
diff --git a/compile_tests/no_display_no_impl.stderr b/compile_tests/no_display_no_impl.stderr
index a6283f5..ae61315 100644
--- a/compile_tests/no_display_no_impl.stderr
+++ b/compile_tests/no_display_no_impl.stderr
@@ -1,11 +1,14 @@
 error[E0277]: `Error` doesn't implement `std::fmt::Display`
- --> compile_tests/no_display_no_impl.rs:1:17
+ --> compile_tests/no_display_no_impl.rs:3:17
   |
-1 | #[derive(Debug, onlyerror::Error)]
+3 | #[derive(Debug, onlyerror::Error)]
   |                 ^^^^^^^^^^^^^^^^ `Error` cannot be formatted with the default formatter
   |
   = help: the trait `std::fmt::Display` is not implemented for `Error`
   = note: in format strings you may be able to use `{:?}` (or {:#?} for pretty-print) instead
 note: required by a bound in `std::error::Error`
  --> $RUST/core/src/error.rs
+  |
+  | pub trait Error: Debug + Display {
+  |                          ^^^^^^^ required by this bound in `Error`
   = note: this error originates in the derive macro `onlyerror::Error` (in Nightly builds, run with -Z macro-backtrace for more info)
diff --git a/compile_tests/one_comment.rs b/compile_tests/one_comment.rs
index 0944401..f4ee70c 100644
--- a/compile_tests/one_comment.rs
+++ b/compile_tests/one_comment.rs
@@ -1,8 +1,9 @@
+#![allow(dead_code)]
+
 #[derive(Debug, onlyerror::Error)]
 enum Error {
-       /// One
-       One,
+    /// One
+    One,
 }

 fn main() {}
-
diff --git a/compile_tests/one_param.rs b/compile_tests/one_param.rs
index 85951f5..dd90b36 100644
--- a/compile_tests/one_param.rs
+++ b/compile_tests/one_param.rs
@@ -1,3 +1,5 @@
+#![allow(dead_code)]
+
 #[derive(Debug, onlyerror::Error)]
 enum Error {
     #[error("One")]
diff --git a/src/parser.rs b/src/parser.rs
index 7162bc6..f39c9c9 100644
--- a/src/parser.rs
+++ b/src/parser.rs
@@ -67,8 +67,7 @@ impl Error {
                 variants,
                 no_display: attributes
                     .into_iter()
-                    .find(|attr| attr.name.to_string() == "no_display")
-                    .is_some(),
+                    .any(|attr| attr.name.to_string() == "no_display"),
             }),
             tree => Err(spanned_error("Unexpected token", tree.as_span())),
         }
@@ -137,7 +136,7 @@ impl Variant {
             ty
         } else {
             // Skip everything before ','
-            while let Err(_) = input.expect_punct(',') {}
+            while input.expect_punct(',').is_err() {}
             VariantType::Unit
         };

```
</details>

**Comment by Mifom** [comment]
> I have one test failure on my end (using both stable and nightly). It appears to be some additional hint included in the error message. There are also a number of warnings from unused enum variants. Those should be silenced with `#![allow(dead_code)]`. And finally some clippy errors. I'll start the CI pipeline and you can see what it catches.
> This is my patch so far to clean those up...

I've done all except additional hint as it can be specific to your configuration. I've tried stable and nightly - no additional hint. So let's look what will be in CI and I'll apply changes to pass it.

**Comment by parasyte** [comment]
I think I see what is happening here. It's probably worth moving ui tests to only one of the supported compilers in the CI build matrix. Cf. https://github.com/dtolnay/trybuild/issues/211

The proposed solution is targeting only the nightly compiler with conditional compilation on the top-level test: https://github.com/dtolnay/ref-cast/blob/f7b0b87cacb4cc0866a07150205e6d35e634fb23/tests/compiletest.rs#L1

I don't run tests on nightly in CI because I have found them to be way too unstable. But [`rustversion`](https://docs.rs/rustversion) knows how to query the beta compiler, so that should work.

**Comment by Mifom** [comment]
As the info differs between stable/beta and 1.62.0 it's obviously is the error output problem. I've found minimal compiler version that have similar output - 1.68.0. So simple solution is bump MSRV to this version. But I don't think it's a good idea.

Test errors only in one version lokks like much better solution, I'll change code into that.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -63,9 +63,7 @@ jobs:
     strategy:
       matrix:
         rust:
-          - stable
           - beta
-          - 1.62.0
```
Comment (parasyte) [inline .github/workflows/ci.yml]: I don't think this will be acceptable, if I'm honest. The conditional compile attribute seems like a better long-term solution. Ultimately, I _do_ want to run unit tests on this entire matrix, even if there are none at present.

Comment (Mifom) [reply .github/workflows/ci.yml]: I don't know many about Github CI config, so is this going to work:
```yaml
    name: Test
    runs-on: ubuntu-latest
    needs: [checks, lints]
    strategy:
      matrix:
        rust:
          - stable
          - beta
          - 1.62.0
    if: matrix.rust == "beta"
    steps:
...

```

Comment (parasyte) [reply .github/workflows/ci.yml]: No, I believe it's more a matter of using an attribute like this in the test: https://github.com/dtolnay/ref-cast/blob/f7b0b87cacb4cc0866a07150205e6d35e634fb23/tests/compiletest.rs#L1 replacing `nightly` with `beta`. The CI stuff can be left as it was.

Comment (Mifom) [reply .github/workflows/ci.yml]: With `rustversion` crate we can set `no_display_no_impl` test to run only since 1.68.0, which is the version I've found to be first with such syntax as stable and beta. So all tests will run in all toolchains and `no_display_no_impl` only in stable/beta

**Review by parasyte — APPROVED** [review]
Thank you!! We will use this for a while and see how it goes.
